Unary Complement execution has different results when the parameters are different

ecwall opened the pull request:
fixes https://github.com/cockroachdb/cockroach/issues/74493

Release note (sql change): Return ambiguous unary operator error for ambiguous input
like ~'1' which can be interpreted as an integer (resulting in -2) or a bit string
(resulting in 0).

Release justification: Improves a confusing error message saying that an operator is
invalid instead of ambiguous.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76943)
<!-- Reviewable:end -->

otan commented:
the tests that fail will be interesting o_o can you comment on why commenting that out fixed things?

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76943)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/sem/tree/overload.go, line 746 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76943#-Mwc7HE7Bb2C0Taz2oFT:-Mwc7HE7Bb2C0Taz2oFU:bqqhfvk) ([raw file](https://github.com/cockroachdb/cockroach/blob/0de95ed22b0c0a2935a7089e097143cb9c9cd858/pkg/sql/sem/tree/overload.go#L746)):*
> ```Go
> 				})
> 		}
> 		if ok, typedExprs, fn, err := checkReturn(ctx, semaCtx, &s); ok {
> ```

`checkReturn` was short circuiting for `'2'` because there was only 1 overload after this filtering above
```
		for _, i := range s.constIdxs {
			constExpr := exprs[i].(Constant)
			s.overloadIdxs = filterOverloads(s.overloads, s.overloadIdxs,
				func(o overloadImpl) bool {
					semaCtx := MakeSemaContext()
					_, err := constExpr.ResolveAsType(ctx, &semaCtx, o.params().GetAt(i))
					return err == nil
				})
		}
```  
This appears to be due to a check to convert the string into a date, but I am not sure why `'1'` is a valid date but `'2'` is not so I am not sure if there are other issues.

This was preventing further filtering from happening below for `'2'` which resulted in 1 overload instead of 0.

1 level up in `func (expr *UnaryExpr) TypeCheck(...)` the 0 overloads are correctly returning an error, but 1 overload is resulting in a result being returned.
```
	if len(fns) != 1 {
		var desStr string
		if desired.Family() != types.AnyFamily {
			desStr = fmt.Sprintf(" (desired <%s>)", desired)
		}
		sig := fmt.Sprintf("%s <%s>%s", expr.Operator, exprReturn, desStr)
		if len(fns) == 0 {
			return nil,
				pgerror.Newf(pgcode.InvalidParameterValue, unsupportedUnaryOpErrFmt, sig)
		}
		fnsStr := formatCandidates(expr.Operator.String(), fns)
		err = pgerror.Newf(pgcode.AmbiguousFunction, ambiguousUnaryOpErrFmt, sig)
		err = errors.WithHintf(err, candidatesHintFmt, fnsStr)
		return nil, err
	}
```

Will this affect other code that calls `typeCheckOverloadedExprs(...)`? Should all short circuits be removed?
```



<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76943)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/sem/tree/overload.go, line 746 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76943#-Mwc7HE7Bb2C0Taz2oFT:-MwcT0VjBFH0bOIOTkMy:b-yobek0) ([raw file](https://github.com/cockroachdb/cockroach/blob/0de95ed22b0c0a2935a7089e097143cb9c9cd858/pkg/sql/sem/tree/overload.go#L746)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

> Will this affect other code that calls typeCheckOverloadedExprs(...)? 

yeah. looks like the test failing here suggest that this check was necessary for certain checks.

> This appears to be due to a check to convert the string into a date,

that doesn't seem right. otherwise, i'd expect this to work:

```
root@127.0.0.1:26257/defaultdb> select '1'::date;
ERROR: parsing as type date: missing required date fields
SQLSTATE: 22007
DETAIL: Wanted: [ Year Day Era Hour Minute Second Nanos Meridian TZHour TZMinute TZSecond ]
Already found in input: [ Month ]
root@127.0.0.1:26257/defaultdb> select '2'::date;
ERROR: parsing as type date: missing required date fields
SQLSTATE: 22007
DETAIL: Wanted: [ Year Day Era Hour Minute Second Nanos Meridian TZHour TZMinute TZSecond ]
Already found in input: [ Month ]
```

what does it think `2` is when we return early is, and why doesn't that result in an error message?
</blockquote></details>

Looking at it further, it is actually not related to a date/time because `ptCtx` is being ignored in the taken branch inside `ParseAndRequireString`
```
		ptCtx := simpleParseTimeContext{
			// We can return any time, but not the zero value - it causes an error when
			// parsing "yesterday".
			RelativeParseTime: time.Date(2000, time.January, 2, 3, 4, 5, 0, time.UTC),
		}
		if semaCtx != nil {
			ptCtx.DateStyle = semaCtx.DateStyle
			ptCtx.IntervalStyle = semaCtx.IntervalStyle
		}
		val, dependsOnContext, err := ParseAndRequireString(typ, expr.s, ptCtx)
```

```
func ParseAndRequireString(
	t *types.T, s string, ctx ParseTimeContext,
) (d Datum, dependsOnContext bool, err error) {
	switch t.Family() {
	case types.ArrayFamily:
		d, dependsOnContext, err = ParseDArrayFromString(ctx, s, t.ArrayContents())
	case types.BitFamily:
		r, err := ParseDBitArray(s)
		if err != nil {
			return nil, false, err
		}
		d = formatBitArrayToType(r, t)
```

The difference still stands though that the reason `'1'` and `'2'` are different is because of an implicit conversion to bit array (1 can be converted which prevents a short circuit thereby allowing later filtering to 0) in cockroach that is not in postgres combined with the early return logic for having 1 overload.


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/sql/sem/tree/overload.go, line 746 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76943#-Mwc7HE7Bb2C0Taz2oFT:-MwhUYrJ5qTjZQ1U0_xp:b-7fy8vt) ([raw file](https://github.com/cockroachdb/cockroach/blob/0de95ed22b0c0a2935a7089e097143cb9c9cd858/pkg/sql/sem/tree/overload.go#L746)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

Looking at it further, it is actually not related to a date/time because `ptCtx` is being ignored in the taken branch inside `ParseAndRequireString`
```
		ptCtx := simpleParseTimeContext{
			// We can return any time, but not the zero value - it causes an error when
			// parsing "yesterday".
			RelativeParseTime: time.Date(2000, time.January, 2, 3, 4, 5, 0, time.UTC),
		}
		if semaCtx != nil {
			ptCtx.DateStyle = semaCtx.DateStyle
			ptCtx.IntervalStyle = semaCtx.IntervalStyle
		}
		val, dependsOnContext, err := ParseAndRequireString(typ, expr.s, ptCtx)
```

```
func ParseAndRequireString(
	t *types.T, s string, ctx ParseTimeContext,
) (d Datum, dependsOnContext bool, err error) {
	switch t.Family() {
	case types.ArrayFamily:
		d, dependsOnContext, err = ParseDArrayFromString(ctx, s, t.ArrayContents())
	case types.BitFamily:
		r, err := ParseDBitArray(s)
		if err != nil {
			return nil, false, err
		}
		d = formatBitArrayToType(r, t)
```

The difference still stands though that the reason `'1'` and `'2'` are different is because of an implicit conversion to bit array (1 can be converted which prevents a short circuit thereby allowing later filtering to 0) in cockroach that is not in postgres combined with the early return logic for having 1 overload.
</blockquote></details>

another case of our type inference system not being great :'(


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76943)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/sem/tree/overload.go, line 746 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76943#-Mwc7HE7Bb2C0Taz2oFT:-Mx63yie4d5vSFONddgv:byod12c) ([raw file](https://github.com/cockroachdb/cockroach/blob/0de95ed22b0c0a2935a7089e097143cb9c9cd858/pkg/sql/sem/tree/overload.go#L746)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

another case of our type inference system not being great :'(
</blockquote></details>

I'm now trying a new strategy: instead of preventing all strings from being implicitly cast, it now will display an `ambiguous unary operator` error (which includes a list of the ambiguities) instead of an `unsupported unary operator` error (which gave no additional info).


<!-- Sent from Reviewable.io -->

otan reviewed (approved):
nice

rafiss commented:
thanks for digging though all this! i like this approach

ecwall committed:
Unary Complement execution has different results when the parameters are
different

fixes https://github.com/cockroachdb/cockroach/issues/74493

Release note (sql change): Return ambiguous unary operator error for ambiguous
input like ~'1' which can be interpreted as an integer (resulting in -2) or a
bit string (resulting in 0).

Release justification: Improves a confusing error message saying that an
operator is invalid instead of ambiguous.

ecwall commented:
bors r=otan
